profiling/src/profiling/interrupts.rs

```diff
@@ -22,39 +26,194 @@ impl std::fmt::Display for VmInterrupt {
+            sync_sender,
+            join_handle: super::thread_utils::spawn(Self::THREAD_NAME, move || {
+                let wall_timer = crossbeam_channel::tick(super::WALL_TIME_PERIOD);
+
+                loop {
+                    // The if/else branches should be the same, except that
+                    // the code path which recv's ticker messages should not
+                    // be present on the branch that handles the case when PHP
+                    // isn't serving a request.
+                    let active_interrupts = !vm_interrupts.lock().unwrap().is_empty();
+                    if active_interrupts {
+                        crossbeam_channel::select! {
+                            recv(async_receiver) -> message => match message {
+                                // This message is just to wake the thread up
+                                // so it can sync `active_interrupts`.
+                                Ok(AsyncMessage::Wake) => {},
+                                Err(err) => {
+                                    log::warn!("{err}");
+                                    break;
+                                },
+                            },
+
+                            recv(sync_receiver) -> message => match message {
+                                Ok(SyncMessage::Pause) => {
+                                    // First, wait for every thread to finish what
+                                    // they are currently doing.
+                                    barrier.wait();
+                                    // Then, wait for the fork to be completed.
+                                    barrier.wait();
+                                },
+
+                                Ok(SyncMessage::Shutdown) => break,
+                                Err(err) => {
+                                    log::warn!("{err}");
+                                    break;
+                                },
+                            },
+
+                            recv(wall_timer) -> message => match message {
+                                Ok(_) => {
+                                    let vm_interrupts = vm_interrupts.lock().unwrap();
+                                    vm_interrupts.iter().for_each(|obj| unsafe {
+                                        (*obj.interrupt_count_ptr).fetch_add(1, Ordering::SeqCst);
+                                        (*obj.engine_ptr).store(true, Ordering::SeqCst);
+                                    });
+                                },
+
+                                Err(err) => {
+                                    log::warn!("{err}");
+                                    break;
+                                },
+                            }
+                        }
+                    } else {
+                        // It's not great that we duplicate this code, but it
+                        // was the only way I could get it to compile.
```

Comment:
Isn't it possible to do `if active_interrupts {.recv(wall_timer) } else { Future::pending() }` in that arm, so that it's never entered in the case interrupts are disabled?

Reply:
This is inside of a crossbeam macro, so I'm not sure. I will poke around a bit here again to see if I can come up with something better.

Reply:
I paired with Colin O'Brien and he noticed a trick in the crossbeam docs I didn't. We can select on a `crossbeam_channel::never` when we're on an iteration which shouldn't use the timer.